SingularityExecutor/src/main/java/com/hubspot/singularity/executor/task/SingularityExecutorTaskProcessCallable.java

```diff
@@ -115,4 +131,84 @@ private boolean runHealthcheck(Process process) {
       return false;
     }
   }
+
+  private boolean runHealthcheckV2(Process process, HealthcheckOptions maybeOptions) {
+    String taskAppDirectory = task.getTaskDefinition().getTaskAppDirectory();
+    File configPath = Paths.get(taskAppDirectory, HEALTHCHECK_CONFIG_FILE_PATH).toFile();
+    LOG.debug("Getting healthcheck config file path at {}", configPath);
+
+    HealthchecksV2 healthchecks;
+    try {
+      healthchecks = objectMapper.readValue(configPath, HealthchecksV2.class);
+    } catch (IOException e) {
+      executorUtils.sendStatusUpdate(task.getDriver(), task.getTaskInfo().getTaskId(), TaskState.TASK_FAILED, String.format("Failed to run healthchecks: {}", e.getCause()), task.getLog());
+      return false;
+    }
+
+    // Start all the futures, ordered by max possibe lifespan
+    List<CompletableFuture<HealthcheckResult>> futures = healthchecks.getHealthchecks().stream()
+        .filter(HealthcheckConfig::isEnabled)
+        .sorted(Comparator.comparingInt(config -> config.getTimeout()))
+        .map(healthcheckConfig -> startHealthcheck(process, healthcheckConfig))
+        .collect(Collectors.toList());
+
+    List<HealthcheckResult> results = futures.stream().map(CompletableFuture::join).collect(Collectors.toList());
+
+    for (HealthcheckResult result : results) {
+      HealthCheckStatus status = result.getStatus();
+      if (status != HealthCheckStatus.PASSED) {
+        executorUtils.sendStatusUpdate(task.getDriver(), task.getTaskInfo().getTaskId(), TaskState.TASK_FAILED, String.format("Healthcheck failed: {}", result.getMessage()), task.getLog());
+        return false;
+      }
+    }
+    executorUtils.sendStatusUpdate(task.getDriver(), task.getTaskInfo().getTaskId(), Protos.TaskState.TASK_RUNNING, String.format("Task running process %s.", getCurrentProcessToString()), task.getLog());
+    return true;
```

Comment:
My biggest concern here is that any given healthcheck
- doesn't prevent another from running
- doesn't prevent early exit due to an earlier failing healthcheck

Reply:
I think this does this, but definitely extra eyes here would be good